pkg/cloudprovider/providers/aws/aws.go

```diff
@@ -1894,31 +1901,44 @@ func (s *AWSCloud) listPublicSubnetIDsinVPC(vpcId string) ([]string, error) {
 	return subnetIds, nil
 }
 
-func isSubnetPublic(rt []*ec2.RouteTable, subnetID string) bool {
+func isSubnetPublic(rt []*ec2.RouteTable, subnetID string) (bool, error) {
+	var mainVpcTable *ec2.RouteTable
 	for _, table := range rt {
-		var found bool
 		for _, assoc := range table.Associations {
-			if aws.StringValue(assoc.SubnetId) == subnetID {
-				found = true
+			if aws.BoolValue(assoc.Main) == true {
+				mainVpcTable = table
 				break
 			}
 		}
-		if !found {
-			continue
-		}
-		for _, route := range table.Routes {
-			// There is no direct way in the AWS API to determine if a subnet is public or private.
-			// A public subnet is one which has an internet gateway route
-			// we look for the gatewayId and make sure it has the prefix of igw to differentiate
-			// from the default in-subnet route which is called "local"
-			// or other virtual gateway (starting with vgv)
-			// or vpc peering connections (starting with pcx).
-			if strings.HasPrefix(aws.StringValue(route.GatewayId), "igw") {
-				return true
+	}
+
+	if mainVpcTable == nil {
+		return false, fmt.Errorf("Could not locate main routing table")
+	}
+
+	subnetTable := mainVpcTable
```

Comment:
Main routing table is required for a VPC afaik. Something would be bad if it was missing. Hmm but we could still recover I suppose if it finds the subnet ids - so I see your point.